File: solutions/python/ocr-numbers/1/ocr_numbers.py
def convert(input_grid):
    numbers = [
        [" _ ", "| |", "|_|", "   "],
        ["   ", "  |", "  |", "   "],
        [" _ ", " _|", "|_ ", "   "],
        [" _ ", " _|", " _|", "   "],
        ["   ", "|_|", "  |", "   "],
        [" _ ", "|_ ", " _|", "   "],
        [" _ ", "|_ ", "|_|", "   "],
        [" _ ", "  |", "  |", "   "],
        [" _ ", "|_|", "|_|", "   "],
        [" _ ", "|_|", " _|", "   "]
    ]
    if len(input_grid) % 4 != 0:
        raise ValueError("Number of input lines is not a multiple of four")
    if any(len(input_column) % 3 != 0 for input_column in input_grid):
        raise ValueError("Number of input columns is not a multiple of three")
    result = []
    for i in range(0, len(input_grid), 4):
        current = ""
        for j in range(0, len(input_grid[0]), 3):
            element = [input_grid[k][j:j+3] for k in range(i, i+4)]
            current += str(numbers.index(element)) if element in numbers else "?"
        result.append(current)
    return ",".join(result)

File: solutions/javascript/ocr-numbers/1/ocr-numbers.js
//
// This is only a SKELETON file for the 'OCR Numbers' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const convert = (sentence) => {
  const numbers = [
    ' _ | ||_|   ',
    '     |  |   ',
    ' _  _||_    ',
    ' _  _| _|   ',
    '   |_|  |   ',
    ' _ |_  _|   ',
    ' _ |_ |_|   ',
    ' _   |  |   ',
    ' _ |_||_|   ',
    ' _ |_| _|   '
  ], result = [];
  sentence = sentence.split("\n");
  for(let i = 0; i < sentence.length; i += 4) {
    let current = "";
    for(let j = 0; j < sentence[i].length; j += 3) {
      let number = "";
      for(let k = i; k < i + 4; k++) 
        number += sentence[k].slice(j, j + 3);
      let idx = numbers.indexOf(number);
      current += idx != -1 ? idx.toString() : '?';
    }
    result.push(current);
  }
  return result.join(',');
};
